corehq/apps/users/user_data.py

```diff
@@ -63,14 +64,19 @@ def save(self):
 
     @property
     def _provided_by_system(self):
-        return {
+        provided_data = {
             **(self.profile.fields if self.profile else {}),
             PROFILE_SLUG: self.profile_id or '',
             COMMCARE_PROJECT: self.domain,
-            COMMCARE_LOCATION_ID: self._couch_user.location_id,
-            COMMCARE_LOCATION_IDS: self._couch_user.location_ids,
-            COMMCARE_PRIMARY_CASE_SHARING_ID: self._couch_user.location_id,
         }
+        if getattr(self._couch_user, 'location_id', None):
```

Comment:
Here I want to keep the behavior remain unchanged. Before I make this change, if no location is assigned to the user, there won't be location info in user_data. The newly added test also ensures my changes won't change the content in user_data, it only changes where it comes from.

Reply:
Hi Ethan, yes, I would love to always include these fields. Can you point me to the code where we access the location info in user data. I'm just worried if there will be some check like: if `location_id` exist in `user_data`, then do something with the value of `location_id`. If the value is "", then that part of code might error.

Reply:
Unfortunately that's what makes this so difficult to reason about - it isn't used in our code, it's used in app logic.  The restore file will contain something like
```xml
  <Registration xmlns="http://openrosa.org/user/registration">
    <username>ethan</username>
    <password>myhashedpassword</password>
    <uuid>myuserid</uuid>
    <date>2024-07-11</date>
    <user_data>
      <data key="commcare_first_name">Ethan</data>
      <data key="commcare_last_name">Sloergel</data>
      <data key="commcare_location_id">52f34e6469b240e6b8bdda7982df2cdd</data>
      <data key="commcare_location_ids">52f34e6469b240e6b8bdda7982df2cdd cab938fece104b888b558cc0b39e96d2</data>
      <data key="commcare_phone_number">9042410933</data>
      <data key="commcare_primary_case_sharing_id">52f34e6469b240e6b8bdda7982df2cdd</data>
      <data key="commcare_profile">2954</data>
      <data key="commcare_project">esoergel</data>
      <data key="commcare_user_type">commcare</data>
      <data key="favorite_color">purple</data>
      <data key="foo">baz</data>
      <data key="government_id"/>
      <data key="yearbook_quote">Not all who wander are lost</data>
    </user_data>
  </Registration>
```
That will be referenced in applications with expressions like
```
instance('commcaresession')/session/user/data/commcare_location_id
```
If that expression references a key that's not present, say `this_isnt_real`, it will raise an error like this:
```
Calculation Error: Error in calculation for /data/missing_field Logic references instance(commcaresession)/session/user/data/this_isnt_real which is not a valid question or value.
```
Whereas referencing a key that is present, but empty (like `government_id`) will return the empty string.

In practice, I think app builder often guard against this field being missing by doing something like this:
```
if(count(instance('commcaresession')/session/user/data/commcare_phone_number) > 0, instance('commcaresession')/session/user/data/commcare_phone_number, "missing")
```
That conditional will behave differently if a field is empty than if it's missing.  That probably means this change is too risky to be warranted, though it is a frustrating situation for everyone, especially app builders.

Reply:
Thank you! Yeah I remember we talked about it when we're having the call. That's why I want to make sure I didn't change any behavior. Btw, I cannot find Matt's PR that saving all user fields, can you give me a link? If we all agree it is risky, can we say we don't need to change my current implementation?

Reply:
Matt's PR is here: https://github.com/dimagi/commcare-hq/pull/34082
Yeah I think it's appropriate to not change behavior here